metrics/multilabelmap.go

```diff
@@ -0,0 +1,229 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package metrics
+
+import (
+	"expvar"
+	"fmt"
+	"io"
+	"reflect"
+	"slices"
+	"strings"
+	"sync"
+)
+
+// Comparable is the type required by MultiLabelMap map keys: something that can
+// be a map key (comparable) but also sorted. The Compare methods for types
+// must be written by hand; no magic is is provided by MultiLabelMap (yet).
+type Comparable[T any] interface {
+	comparable
+	Compare(T) int
+}
+
+// MultiLabelMap is a struct-value-to-Var map variable that satisfies the
+// [expvar.Var] interface but also allows for multiple Prometheus labels to be
+// associated with each value.
+//
+// T must be a struct type with only string fields. The struct field names
+// (lowercased) are used as the labels, unless a "prom" struct tag is present.
+// The struct fields must all be strings, and the string values must be valid
+// Prometheus label values without requiring quoting.
+type MultiLabelMap[T Comparable[T]] struct {
+	m         sync.Map  // map[T]Var
+	tp        sync.Pool // of *T
+	keysMu    sync.RWMutex
+	keys      []T                                                           // sorted
+	writeProm func(w io.Writer, name string, k reflect.Value, v expvar.Var) // nil until set
+}
+
+func genWriteFunc(t reflect.Type) func(w io.Writer, name string, k reflect.Value, v expvar.Var) {
+	k := t.Kind()
+	if k != reflect.Struct {
+		panic(fmt.Sprintf("MultiLabelMap must use keys of type struct; got %v", t))
+	}
+	var labels []string
+	for i := 0; i < t.NumField(); i++ {
+		ft := t.Field(i)
+		label := ft.Tag.Get("prom")
+		if label == "" {
+			label = strings.ToLower(ft.Name)
+		}
+		labels = append(labels, label)
+	}
+	return func(w io.Writer, name string, k reflect.Value, v expvar.Var) {
+		io.WriteString(w, name)
+		io.WriteString(w, "{")
+		for i, label := range labels {
+			if i > 0 {
+				fmt.Fprint(w, ",")
+			}
+			s := k.Field(i).String()
+			io.WriteString(w, label)
```

Comment:
We need to make sure that either label names only contain [valid characters](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels), or run them through something like [mapInvalidMetricRunes](https://github.com/tailscale/tailscale/blob/6d90966c1fc097c7738f81cf8a14538482221233/tsweb/varz/varz.go#L94) here. Note that label names cannot contain `:` (but metric names can).

Reply:
I just added that mapInvalidMetricRunes the other day. I remember :)